src/platform/w3dengine/client/w3dbridgebuffer.cpp

```diff
@@ -0,0 +1,788 @@
+{
+    if (mesh == nullptr) {
+        return 0;
+    }
+
+    int count = mesh->Peek_Model()->Get_Polygon_Count();
+    const TriIndex *polys = mesh->Peek_Model()->Get_Polygon_Array();
+
+    if (cur_index + 3 * count + 6 >= W3DBridgeBuffer::MAX_BRIDGE_INDEX) {
+        return 0;
+    }
+
+    unsigned short *index = &destination_ib[cur_index];
+
+    for (int i = 0; i < count; i++) {
+        *index = polys[i].I + vertex_offset;
+        index++;
+        *index = polys[i].J + vertex_offset;
+        index++;
+        *index = polys[i].K + vertex_offset;
+        index++;
+    }
+
+    return 3 * count;
+}
+
+int W3DBridge::Get_Model_Vertices(VertexFormatXYZNDUV1 *destination_vb,
+    int cur_vertex,
+    float x_offset,
+    Vector3 &vec,
+    Vector3 &vec_normal,
+    Vector3 &vec_z,
+    Vector3 &offset,
+    Matrix3D const &mtx,
+    MeshClass *mesh,
+    RefMultiListIterator<RenderObjClass> *lights)
+{
+    if (mesh == nullptr) {
+        return 0;
+    }
+
+    int count = mesh->Peek_Model()->Get_Vertex_Count();
+    Vector3 *verts = mesh->Peek_Model()->Get_Vertex_Array();
+    const Vector3 *normals = mesh->Peek_Model()->Get_Vertex_Normal_Array();
+
+    if (cur_vertex + count + 2 >= W3DBridgeBuffer::MAX_BRIDGE_VERTEX) {
+        return 0;
+    }
+
+    Vector3 light[LIGHT_COUNT];
+
+    for (int i = 0; i < g_theWriteableGlobalData->m_numberGlobalLights; i++) {
+        light[i].Set(-g_theWriteableGlobalData->m_terrainLightPos[i].x,
+            g_theWriteableGlobalData->m_terrainLightPos[i].y,
+            g_theWriteableGlobalData->m_terrainLightPos[i].z);
+        light[i].Normalize();
+    }
+
+    const Vector2 *uvs = mesh->Peek_Model()->Get_UV_Array_By_Index(0);
+    VertexFormatXYZNDUV1 *vertex = &destination_vb[cur_vertex];
+
+    for (int i = 0; i < count; i++) {
+        Vector3 pos;
+        Vector3 out;
+        Matrix3D::Transform_Vector(mtx, verts[i], &out);
+        pos = ((out.X + x_offset) * vec) + (out.Y * vec_normal) + (out.Z * vec_z);
+        pos.X += m_start.X;
+        pos.Y += m_start.Y;
+        pos.Z += m_start.Z;
+        vertex->x = pos.X;
+        vertex->y = pos.Y;
+        vertex->z = pos.Z;
+
+        VertexFormatXYZDUV2 vb;
+        vb.x = pos.X;
+        vb.y = pos.Y;
+        vb.z = pos.Z;
+        Vector3 normal;
+        Matrix3D::Rotate_Vector(mtx, normals[i], &normal);
+        normal = (normal.X * vec) + (normal.Y * vec_normal) + (normal.Z * vec_z);
+        normal.Normalize();
+
+        g_theTerrainRenderObject->Do_The_Light(&vb, light, &normal, nullptr, 1, false);
+        vertex->nx = 0.0f;
+        vertex->ny = 0.0f;
+        vertex->nz = 0.0f;
+
+        vertex->diffuse = vb.diffuse | 0xFF000000;
+        vertex->u1 = uvs[i].X;
+        vertex->v1 = uvs[i].Y;
+        vertex++;
+    }
+
+    return count;
+}
+
+void W3DBridge::Get_Indices_N_Vertices(unsigned short *destination_ib,

```

Comment:
this function can be nearly completely flattened by adding return on each assert as it doesn't seem to branch out anywhere else on any case

Reply:
fixed.